File: Data Structure/Linked Lists/Circular Linked List/CircularLinkedList.py

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
    

class CircularLinkedList:
    def __init__(self):
        self.tail =  None


    def traverse(self):
        if self.tail:
            print("tail= ", self.tail.data)
            print("traverse= ", end='')
            head =  self.tail.next
            if(self.tail.next != None):
                while True:
                    print(head.data, " ", end='')
                    head = head.next
                    if head == self.tail.next:
                        break
            else:
                print(self.tail.data)
        
    

    def insertAtBegining(self, data):
        temp =  Node(data)
        if(self.tail!=None):
            if((self.tail).next!=None):
                temp.next = (self.tail).next
            else:
                temp.next = self.tail
            (self.tail).next = temp
        else:
            self.tail = temp
        

    def insertAtEnd(self, data):
        temp = Node(data)
        if(self.tail != None):
            temp.next = (self.tail).next
            (self.tail).next = temp
            self.tail = temp
        else:
            self.tail = temp
    

    def insertAtPosition(self, position, data):
        temp = Node(data)
        if(self.tail != None):
            positionPointer = self.tail.next
            i = 1
            while(i<position and positionPointer != None and positionPointer.next!=None):
                positionPointer = positionPointer.next
                i = i+1
            temp.next = positionPointer.next
            positionPointer.next = temp
            if(positionPointer == self.tail):
                self.tail = temp
        else:
            self.tail = temp
        
    

    def deleteAtBeginning(self):
        if((self.tail) == None):
            print("No Node")
            return

        if((self.tail).next == None):
            self.tail = None
            return

        if((self.tail).next.next == None):
            (self.tail).next = None
            return
        
        (self.tail).next = (self.tail).next.next
        return
    

    def deleteAtEnd(self):
        temp = self.tail
        while(temp!=None and temp.next!=None and temp.next!=self.tail):
            temp = temp.next
        
        if(temp==None or temp.next==None):
            self.tail = None
            return
        
        temp.next = (self.tail).next
        self.tail = temp 
    

    def deleteAtPosition(self, position):
        temp = (self.tail).next

        i = 1
        while(i<position and temp!=None and temp.next!=None and temp.next!=self.tail):    
            temp = temp.next
            i=i+1
        
        if(temp==None or temp.next==None):
            self.tail = None
            return
        
        if(temp.next == self.tail):
            temp.next = temp.next.next
            (self.tail) = temp
            return
        
        temp.next = temp.next.next
        
    

if __name__ == "__main__":
    cll = CircularLinkedList()
    cll.traverse()
    cll.insertAtBegining('D')
    cll.insertAtBegining('C')
    cll.insertAtBegining('B')
    cll.insertAtBegining('A')
    cll.traverse()
    cll.insertAtEnd('F')
    cll.insertAtPosition(4, 'E')
    cll.insertAtPosition(6, 'G')
    cll.traverse()
    cll.deleteAtBeginning()
    cll.deleteAtEnd()
    # cll.traverse()
    cll.deleteAtPosition(3)
    cll.traverse()

File: Data Structure/Linked Lists/Circular Linked List/CircularLinkedList.cpp
# include<iostream>
using namespace std;
class Node{
    public:
        char data;
        Node* next = nullptr;
    
    public:
        Node(char data1){
            data = data1;
        }
};

void traverse(Node *cll){
    cout << "tail= "<< cll->data << endl;
    cout << "traverse= ";
    Node* head =  cll->next;
    if(cll->next != nullptr){
        do{
            cout << head->data << " ";
            head = head->next;
        }while(head != cll->next);
    }
    else{
        cout << cll->data << endl;
    }
}

void insertAtBegining(Node** cll, char data){
    Node* temp = new Node(data);
    if(cll!=nullptr){
        if((*cll)->next!=nullptr){
            temp->next = (*cll)->next;
        }
        else{
             temp->next = *cll;
        }
        (*cll)->next = temp;
    }
    else{
        *cll = temp;
    }
}

void insertAtEnd(Node** cll, char data){
    Node* temp = new Node(data);
    if(*cll != nullptr){
        temp->next = (*cll)->next;
        (*cll)->next = temp;
        (*cll) = temp;
    }
    else{
        *cll = temp;
    }
}

void insertAtPosition(Node** cll, int position, char data){
    Node* temp = new Node(data);
    if(*cll != nullptr){
        Node* positionPointer = (*cll)->next;
        for(int i = 1; i<position && positionPointer != nullptr && positionPointer->next!=nullptr; i++){
            positionPointer = positionPointer->next;
        }
        temp->next = positionPointer->next;
        positionPointer->next = temp;
        if(positionPointer == *cll){
            *cll = temp;
        }
    }
    else{
        *cll = temp;
    }
}

void deleteAtBeginning(Node** cll){
    if((*cll) == nullptr){
        cout<<"No Node";
        return;
    }
    if((*cll)->next == nullptr){
        *cll = nullptr;
        return;
    }
    if((*cll)->next->next == nullptr){
        (*cll)->next = nullptr;
        return;
    }
    Node* head = (*cll)->next;
    (*cll)->next = (*cll)->next->next;
    delete head;
    return;
}

void deleteAtEnd(Node** cll){
    Node* temp = *cll;
    while(temp!=nullptr && temp->next!=nullptr && temp->next!=*cll){
        temp = temp->next;
    }
    if(temp==nullptr||temp->next==nullptr){
        *cll = nullptr;
        return;
    }
    temp->next = (*cll)->next;
    delete (*cll);
    *cll = temp; 
}

void deleteAtPosition(Node** cll, int position){
    Node* temp = (*cll)->next;
    for(int i = 1; i<position && temp!=nullptr && temp->next!=nullptr && temp->next!=*cll; i++){
        
        temp = temp->next;
        cout <<temp->data;
    }
    if(temp==nullptr||temp->next==nullptr){
        *cll = nullptr;
        return;
    }
    if(temp->next == *cll){
        temp->next = temp->next->next;
        delete *cll;
        (*cll) = temp;
        return;
    }
    temp->next = temp->next->next;
    
}


int main(){
    Node* cll = new Node('D');
    traverse(cll);
    insertAtBegining(&cll, 'C');
    insertAtBegining(&cll, 'B');
    insertAtBegining(&cll, 'A');
    insertAtEnd(&cll, 'F');
    insertAtPosition(&cll, 4, 'E');
    insertAtPosition(&cll, 6, 'G');
    traverse(cll);
    deleteAtBeginning(&cll);
    deleteAtEnd(&cll);
    traverse(cll);

    deleteAtPosition(&cll, 3);
    traverse(cll);
    return 0;
}